resource/graph_node.go

```diff
@@ -518,11 +525,17 @@ func (w *GraphNode) resourceStatus() Status {
 		resName = w.current.Name()
 	}
 
+	var err error
+	if w.state == NodeStateUnhealthy {
+		err = w.lastErr
```

Comment:
Would it be an invariant violation for either of:
* `w.state == Unhealthy and w.lastErr == nil` or
* `w.state != Unhealthy and w.lastErr != nil`

Can we add log line warns for any combinations that shouldn't be allowed here?

Reply:
It should be, I'll add a log and verify that we don't leave a "dangling" error when we transition from the unhealthy state.

Reply:
Double checking, was there a log line for this? `transitionTo` seems unchanged. Maybe there are multiple functions making mutations?

Reply:
> It should be

I misspoke, errors are only cleared when a new healthy resource is "swapped" into the node, so a node can have an error without being in the unhealthy state. Here's my current understanding of the possibilities:

* **A node *can* be in the `Removing` state with an error** because the system marks nodes for removal without clearing the error. It's not clear to me if this is problem, so maybe we don't need a warning here - perhaps `Debug` might be more appropriate if we still want to log something.
* **A node *might* be in the `Configuring` state with an error**, but it's possible that our system filters out unhealthy nodes when they would otherwise be transitioned to `Configuring` - see [this thread](https://github.com/viamrobotics/rdk/pull/4257#discussion_r1712204973) for details. A warning log would be helpful in this case.
* **A node *should never* be `Ready` with an error**. I'm fairly certain this can't happen since nodes are only marked `Ready` when a healthy resource is swapped in. A warning log is certainly warranted here.

Reply:
Forgot to comment on `w.state == Unhealthy and w.lastErr == nil`, this invariant should hold, but I will add a warning nonetheless.

Reply:
Agreed that removing + error is not problematic. Can omit a log line or debug is fine with me.
Also agree on Configuring + error is legal
And agree on ready + error -> warning log.

More relevant to this specific diff, I imagine we should instead just always grab the `err` field -- not only when the node is unhealthy?

Or I'd be fine with not grabbing node.err when node.state == `Ready`. With the expectation that there will be a log line for that case.

Reply:
> More relevant to this specific diff, I imagine we should instead just always grab the err field -- not only when the node is unhealthy?
> 
> Or I'd be fine with not grabbing node.err when node.state == Ready. With the expectation that there will be a log line for that case.

I think option 2 is the way to go, with a commitment to resolving any errors that violate the invariant.

I will update [our API docs](https://github.com/viamrobotics/api/blob/13d281727f811cfcc3979ef83498913dea2b1d63/proto/viam/robot/v1/robot.proto#L413) to reflect the new behavior.